File: java/src/tarehart/rlbot/steps/defense/ThreatAssessor.java
package tarehart.rlbot.steps.defense;

import tarehart.rlbot.AgentInput;
import tarehart.rlbot.input.CarData;
import tarehart.rlbot.intercept.InterceptCalculator;
import tarehart.rlbot.math.VectorUtil;
import tarehart.rlbot.math.vector.Vector3;
import tarehart.rlbot.physics.ArenaModel;
import tarehart.rlbot.physics.BallPath;
import tarehart.rlbot.planning.Goal;
import tarehart.rlbot.planning.GoalUtil;
import tarehart.rlbot.intercept.Intercept;
import tarehart.rlbot.time.Duration;

import java.util.Optional;

public class ThreatAssessor {


    public double measureThreat(AgentInput input, Optional<CarData> enemyCarOption) {

        double enemyPosture = measureEnemyPosture(input, enemyCarOption);
        double enemyInitiative = measureEnemyInitiative(input, enemyCarOption);
        double ballThreat = measureBallThreat(input) *  .3;

        double enemyThreat = enemyPosture > 0 && enemyInitiative > .2 ? 10 : 0;

        return enemyThreat + ballThreat;

    }

    private double measureEnemyInitiative(AgentInput input, Optional<CarData> enemyCarOption) {

        if (!enemyCarOption.isPresent()) {
            return 0;
        }
        CarData enemyCar = enemyCarOption.get();

        Duration simDuration = Duration.Companion.ofSeconds(4);
        BallPath ballPath = ArenaModel.predictBallPath(input);

        CarData myCar = input.getMyCarData();

        Optional<Intercept> myInterceptOption = InterceptCalculator.INSTANCE.getInterceptOpportunityAssumingMaxAccel(myCar, ballPath, myCar.getBoost());
        Optional<Intercept> enemyInterceptOption = InterceptCalculator.INSTANCE.getInterceptOpportunityAssumingMaxAccel(enemyCar, ballPath, enemyCar.getBoost());

        if (!enemyInterceptOption.isPresent()) {
            return 0;
        }

        if (!myInterceptOption.isPresent()) {
            return 3;
        }

        Intercept myIntercept = myInterceptOption.get();
        Intercept enemyIntercept = enemyInterceptOption.get();

        return Duration.Companion.between(myIntercept.getTime(), enemyIntercept.getTime()).getSeconds();
    }

    private double measureEnemyPosture(AgentInput input, Optional<CarData> enemyCarOption) {

        if (!enemyCarOption.isPresent()) {
            return 0;
        }
        CarData enemyCar = enemyCarOption.get();
        
        Goal myGoal = GoalUtil.INSTANCE.getOwnGoal(input.getTeam());
        Vector3 ballToGoal = myGoal.getCenter().minus(input.getBallPosition());

        Vector3 carToBall = input.getBallPosition().minus(enemyCar.getPosition());
        Vector3 rightSideVector = VectorUtil.INSTANCE.project(carToBall, ballToGoal);

        return rightSideVector.magnitude() * Math.signum(rightSideVector.dotProduct(ballToGoal));
    }


    private double measureBallThreat(AgentInput input) {

        CarData car = input.getMyCarData();
        Goal myGoal = GoalUtil.INSTANCE.getOwnGoal(input.getTeam());
        Vector3 ballToGoal = myGoal.getCenter().minus(input.getBallPosition());

        Vector3 ballVelocityTowardGoal = VectorUtil.INSTANCE.project(input.getBallVelocity(), ballToGoal);
        double ballSpeedTowardGoal = ballVelocityTowardGoal.magnitude() * Math.signum(ballVelocityTowardGoal.dotProduct(ballToGoal));

        Vector3 carToBall = input.getBallPosition().minus(car.getPosition());
        Vector3 wrongSideVector = VectorUtil.INSTANCE.project(carToBall, ballToGoal);
        double wrongSidedness = wrongSideVector.magnitude() * Math.signum(wrongSideVector.dotProduct(ballToGoal));

        return ballSpeedTowardGoal + wrongSidedness;
    }

}

File: src/main/java/tarehart/rlbot/steps/defense/ThreatAssessor.kt
package tarehart.rlbot.steps.defense

import tarehart.rlbot.TacticalBundle
import tarehart.rlbot.intercept.strike.JumpHitStrike
import tarehart.rlbot.math.vector.Vector2
import tarehart.rlbot.planning.GoalUtil
import tarehart.rlbot.planning.ZoneUtil
import tarehart.rlbot.time.Duration

class ThreatAssessor {


    companion object {
        fun getThreatReport(bundle: TacticalBundle): ThreatReport {

            val enemyCar = bundle.tacticalSituation.enemyPlayerWithInitiative

            val myGoal = GoalUtil.getOwnGoal(bundle.agentInput.team)
            val ballPosition = bundle.agentInput.ballPosition
            val car = bundle.agentInput.myCarData
            val ballIsBehindUs = myGoal.center.distanceSquared(ballPosition) < myGoal.center.distanceSquared(car.position)

            if (enemyCar?.intercept == null) {
                return ThreatReport(
                        enemyMightBoom = false,
                        enemyShotAligned = false,
                        enemyWinsRace = false,
                        enemyHasBreakaway = false,
                        ballIsBehindUs = ballIsBehindUs,
                        enemyDribbling = false,
                        challengeImminent = false)
            }

            val enemyIntercept = enemyCar.intercept
            val enemyFlatPos = enemyCar.car.position.flatten()
            val enemyToIntercept = enemyIntercept.space.flatten() - enemyFlatPos
            val shotAlignment = Vector2.alignment(enemyFlatPos, enemyIntercept.space.flatten(), myGoal.center.flatten())

            val enemyCarToStrike = enemyIntercept.space - enemyCar.car.position
            val strikeVelocity = enemyCarToStrike.scaledToMagnitude(enemyIntercept.accelSlice.speed)
            val impactSpeed = (strikeVelocity - bundle.agentInput.ballVelocity).magnitude()

            val enemyFlatVel = enemyCar.car.velocity.flatten()
            val enemyDribbling = enemyCar.car.position.distance(bundle.agentInput.ballPosition) < 10 &&
                        enemyFlatVel.distance(bundle.agentInput.ballVelocity.flatten()) < 10

            var challengeImminent = false
            val ourIntercept = bundle.tacticalSituation.expectedContact
            val ourTimeToEnemyIntercept = ourIntercept?.distancePlot?.getMotionUponArrival(car, enemyIntercept.space)?.time
            if (ourIntercept != null && ourTimeToEnemyIntercept != null) {
                // Sometimes our own intercept time is not very important because the enemy is going to get there first
                // and redirect the ball.

                challengeImminent = ourIntercept.space.z < 5 &&
                        bundle.tacticalSituation.ballAdvantage.seconds < 0.3 &&
                        ourTimeToEnemyIntercept < Duration.ofSeconds(1.5) &&
                        enemyIntercept.time.isBefore(car.time.plusSeconds(1.0)) &&
                        enemyFlatVel.normalized().dotProduct(enemyToIntercept.normalized()) > 0.5 &&  // Enemy driving toward intercept
                        shotAlignment > 0.6 // Enemy is a threat to our goal
            }


            return ThreatReport(
                    enemyMightBoom = impactSpeed > 30 && shotAlignment > 0.6 && enemyIntercept.space.z < JumpHitStrike.MAX_BALL_HEIGHT_FOR_JUMP_HIT,
                    enemyShotAligned = shotAlignment > 0.6 && enemyIntercept.space.z < JumpHitStrike.MAX_BALL_HEIGHT_FOR_JUMP_HIT,
                    enemyWinsRace = bundle.tacticalSituation.ballAdvantage.millis < 0,
                    enemyHasBreakaway = ZoneUtil.isEnemyOffensiveBreakaway(car, enemyCar.car, ballPosition),
                    ballIsBehindUs = ballIsBehindUs,
                    enemyDribbling = enemyDribbling,
                    challengeImminent = challengeImminent
            )

        }

    }
}
